libs/cassandra-util/src/Cassandra/Util.hs

```diff
@@ -17,32 +17,100 @@
 
 module Cassandra.Util
   ( defInitCassandra,
+    initCassandraForService,
+    initCassandra,
     Writetime (..),
     writetimeToInt64,
   )
 where
 
-import Cassandra (ClientState, init)
 import Cassandra.CQL
-import Cassandra.Settings (defSettings, setContacts, setKeyspace, setLogger, setPortNumber)
+import Cassandra.Schema
+import Cassandra.Settings (dcFilterPolicyIfConfigured, initialContactsDisco, initialContactsPlain, mkLogger)
 import Data.Aeson
 import Data.Fixed
-import Data.Text (unpack)
+import Data.List.NonEmpty qualified as NE
+import Data.Text (pack, unpack)
 import Data.Time (UTCTime, nominalDiffTimeToSeconds)
 import Data.Time.Clock (secondsToNominalDiffTime)
 import Data.Time.Clock.POSIX
+import Database.CQL.IO
 import Database.CQL.IO.Tinylog qualified as CT
 import Imports hiding (init)
+import OpenSSL.Session qualified as OpenSSL
 import System.Logger qualified as Log
 
-defInitCassandra :: Text -> Text -> Word16 -> Log.Logger -> IO ClientState
-defInitCassandra ks h p lg =
-  init
-    $ setLogger (CT.mkLogger lg)
-      . setPortNumber (fromIntegral p)
-      . setContacts (unpack h) []
-      . setKeyspace (Keyspace ks)
-    $ defSettings
+defInitCassandra :: Text -> Text -> Word16 -> Maybe FilePath -> Log.Logger -> IO ClientState
+defInitCassandra ks h p mbTlsCaPath logger = do
+  let basicCasSettings =
+        setLogger (CT.mkLogger logger)
+          . setPortNumber (fromIntegral p)
+          . setContacts (unpack h) []
+          . setKeyspace (Keyspace ks)
+          . setProtocolVersion V4
+          $ defSettings
+  initCassandra basicCasSettings mbTlsCaPath logger
+
+-- | Create Cassandra `ClientState` ("connection") for a service
+--
+-- Unfortunately, we have to deal with many function arguments here, because
+-- @CassandraOpts@ is defined in @types-common@ which depends on
+-- @cassandra-util@ (this package.)
+initCassandraForService ::
+  Text ->
+  Word16 ->
+  String ->
+  Text ->
+  Maybe FilePath ->
+  Maybe Text ->
+  Maybe Text ->
+  Maybe Int32 ->
+  Log.Logger ->
+  IO ClientState
+initCassandraForService host port serviceName keyspace mbTlsCaPath filterNodesByDatacentre discoUrl mbSchemaVersion logger = do
+  c <-
+    maybe
+      (initialContactsPlain host)
+      (initialContactsDisco ("cassandra_" ++ serviceName) . unpack)
+      discoUrl
```

Comment:
This disco stuff is not used anymore, we should delete it. Perhaps in a separate PR.

Reply:
I've created an issue: https://wearezeta.atlassian.net/browse/WPB-5927